astropy/uncertainty/uncertainty.py

```diff
@@ -0,0 +1,306 @@
+# -*- coding: utf-8 -*-
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+"""
+Distribution class and associated machinery.
+"""
+import numpy as np
+
+from astropy import units as u
+from astropy.utils.misc import isiterable
+
+
+__all__ = ['Uncertainty', 'Variable']
+
+
+# Derivatives of ufuncs relative to their input(s).
+UFUNC_DERIVATIVES = {
+    np.add: (lambda x, y: 1.,
+             lambda x, y: 1.),
+    np.subtract: (lambda x, y: 1.,
+                  lambda x, y: -1.),
+    np.multiply: (lambda x, y: y,
+                  lambda x, y: x),
+    np.true_divide: (lambda x, y: 1/y,
+                     lambda x, y: -x/y**2),
```

Comment:
In terms of performance, `np.square` typically tends to be somewhat faster (5 - 15% in my experience) than using the  power operator to square.  Replacing all the squares here with `np.square` should give you a moderate performance boost.  The same is probably true for `np.power`, but I have not tested this.

Reply:
Numpy makes this optimization internally for `__pow__`: https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/number.c#L507

Reply:
Interesting.  I wasn't aware of this optimization.  I convinced myself that `np.square` was faster by running this test, but I suppose these results are themselves uncertain (ahem): 
```
for p in range(1, 8):
    n = int(10 ** p)
    x = np.random.rand(n)
    t_pow = %timeit -o x ** 2
    t_sqr = %timeit -o np.square(x)
    speedup = np.mean(1 - np.divide(t_sqr.timings, t_pow.timings))
    print('speedup (n = %.0e): %.2f%%' % (n, speedup * 100))
```
The results were: 
```
971 ns ± 63.9 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
911 ns ± 52.9 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
speedup (n = 1e+01): 5.71%
977 ns ± 31.5 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
909 ns ± 23.4 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
speedup (n = 1e+02): 6.84%
1.6 µs ± 45.9 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
1.53 µs ± 73.9 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
speedup (n = 1e+03): 4.19%
5.23 µs ± 407 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
4.81 µs ± 302 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
speedup (n = 1e+04): 7.70%
64.2 µs ± 5.33 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
66 µs ± 7.62 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
speedup (n = 1e+05): -3.82%
1.86 ms ± 181 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
1.86 ms ± 183 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
speedup (n = 1e+06): -0.71%
55.4 ms ± 3.56 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
47.9 ms ± 7.34 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
speedup (n = 1e+07): 12.76%
```

Reply:
Quite large fluctuations, indeed. Overall, my sense would be to leave it for the future - there may be larger advantages to be had by carefully ensuring more complicated operations are done in-place (though some of that is also done automatically in numpy).